packages/gatsby-source-contentful/src/source-nodes.js

```diff
@@ -58,16 +59,22 @@ export async function sourceNodes(
     actions
   const online = await isOnline()
 
-  getNodes().forEach(node => {
-    if (node.internal.owner !== `gatsby-source-contentful`) {
-      return
-    }
-    touchNode(node)
-    if (node?.fields?.localFile) {
-      // Prevent GraphQL type inference from crashing on this property
-      touchNode(getNode(node.fields.localFile))
-    }
-  })
+  // Gatsby only checks if a node has been touched on the first sourcing.
+  // As iterating and touching nodes can grow quite expensive on larger sites with
+  // 1000s of nodes, we'll skip doing this on subsequent sources.
+  if (isFirstSource) {
+    getNodes().forEach(node => {
+      if (node.internal.owner !== `gatsby-source-contentful`) {
+        return
+      }
+      touchNode(node)
+      if (node?.fields?.localFile) {
+        // Prevent GraphQL type inference from crashing on this property
+        touchNode(getNode(node.fields.localFile))
```

Comment:
Doesn't getNode return a promise?

Reply:
It's sync afaik

Reply:
This feels like a bug in Gatsby. Any investigation you did that tells you why you need this?

Reply:
Which part? I only added the check if it's an initial build or not — the bits here around `localFile` I didn't write.

Also just noticing that Tyler was asking about `getNode` instead of `touchNode` 🤦 but luckily it's sync as well 😆

Reply:
Ah, didn't realize that! Perhaps it was async before lmdb 🤔 not sure why I thought that

Reply:
it was sync then too (as it was just pulling the node off a Map) — that LMDB supported sync gets was a huge mark in its favor as we otherwise would have a big breaking change for the framework and source/transformer plugins that all relied on the previous sync `getNode`

Reply:
Good to know :) I guess I was just needlessly treating it like a promise before 😜

Reply:
Node.js is forgiving :-D

Reply:
> Which part? I only added the check if it's an initial build or not — the bits here around `localFile` I didn't write.
> 

Oh I see. I was talking about localFile yes